### ISSUE TITLE
Fatal crash - NullPointerException - Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference

### ISSUE THREAD

**Issue by tingsu** [ankidroid/Anki-Android#6028]
###### Reproduction Steps

The steps were reported by my testing tool (note these steps may not be complete or shortest, but it could lead to the crash). I annotated the action on some key steps, and hope they are useful to help you diagonise the root cause.  I tried to manually reproduce this issue, but it is not easy for me to come up the valid trace. Maybe you folks are more familiar with the code and can identify the root cause more quickly :)

Step 1:
![Selection_216](https://user-images.githubusercontent.com/3156978/79448849-775ad780-7fe2-11ea-9feb-51c6ce91cf52.png)

Step 2 (click ``Empty``):
![Selection_217](https://user-images.githubusercontent.com/3156978/79448871-7fb31280-7fe2-11ea-988b-e1bbd0a86170.png)

Step 3:
![Selection_218](https://user-images.githubusercontent.com/3156978/79448960-a3765880-7fe2-11ea-8948-ee9c3863de67.png)

Step 4:
![Selection_219](https://user-images.githubusercontent.com/3156978/79448969-a83b0c80-7fe2-11ea-9cc9-0981bbed41a8.png)

Step 5:
![Selection_220](https://user-images.githubusercontent.com/3156978/79448982-ac672a00-7fe2-11ea-9676-c9acec319f14.png)

Step 6:
![Selection_221](https://user-images.githubusercontent.com/3156978/79448989-affab100-7fe2-11ea-9ca5-e3a28ac0c081.png)

Step 7:
![Selection_222](https://user-images.githubusercontent.com/3156978/79448994-b1c47480-7fe2-11ea-9241-690e13d04899.png)

Step 8:
![Selection_223](https://user-images.githubusercontent.com/3156978/79448999-b4bf6500-7fe2-11ea-8772-605231ad117d.png)

Step 9  (click the ``undo`` icon):
![Selection_224](https://user-images.githubusercontent.com/3156978/79449005-b721bf00-7fe2-11ea-9519-d7d832a389f5.png)

Step 10:
![Selection_225](https://user-images.githubusercontent.com/3156978/79449008-b8eb8280-7fe2-11ea-8050-eef18ae9bf23.png)

Step 11 (click ``Empty Cards``):
![Selection_226](https://user-images.githubusercontent.com/3156978/79449012-bb4ddc80-7fe2-11ea-8c54-1c8e358970ab.png)

###### Expected Result

No crash

###### Actual Result

Crash happens

###### Debug info

This crash was found in the recent released version 2.9.6 on an Android 6.0 device.

```
 FATAL EXCEPTION: AsyncTask #1
 Process: com.ichi2.anki, PID: 4195
 java.lang.RuntimeException: An error occurred while executing doInBackground()
 	at android.os.AsyncTask$3.done(AsyncTask.java:309)
 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
 	at java.lang.Thread.run(Thread.java:818)
 Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
 	at com.ichi2.libanki.Collection.genCards(Collection.java:761)
 	at com.ichi2.libanki.Collection.genCards(Collection.java:708)
 	at com.ichi2.libanki.Collection.emptyCids(Collection.java:966)
 	at com.ichi2.async.DeckTask.doInBackGroundFindEmptyCards(DeckTask.java:1500)
 	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:346)
 	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:70)
 	at android.os.AsyncTask$2.call(AsyncTask.java:295)
 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
 	... 4 more
```

###### Research
*Enter an [ x ] character to confirm the points below:*

[ X ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ X ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ X ] I have searched for similar existing issues here and on the user forum

**Comment by david-allison-1** [comment]
This is an outstanding report, thank you so much! 

Despite this, I still can't replicate it.

* Is this under the V2 Scheduler?
* How and where were the notes deleted before reaching this state

<details>
<summary>Some relevant code</summary>

Line of the crash
https://github.com/ankidroid/Anki-Android/blob/b06592eae0aa642753073c8c8af192692f4c8ff1/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java#L761

Changing settings rebuilds the deck
https://github.com/ankidroid/Anki-Android/blob/916f55eec5a830d33c2505074d206b63559753cc/AnkiDroid/src/main/java/com/ichi2/anki/FilteredDeckOptions.java#L423-L430
</details>

This happens in the wild as well, so not just an automated testing crash. (Saw a few on 2.8.4 as well)

Samples 

2.9.6:
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/eeea7b9f-ac9a-4a76-b2ab-a868bd7e1f5e
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/5367d828-44d0-4496-b5fe-47d3a65b1fa5
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/7e2a3606-6485-4d3c-83eb-bdc8abb5b693
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/b002f696-fa11-4e19-b297-f3fe7993edf6
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/41d8e299-ff5a-48ec-aba8-4592988f7d08
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/d28e9823-c616-46ba-966e-9d60ab5f980f
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/d28e9823-c616-46ba-966e-9d60ab5f980f

2.9.5:
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/73b43a4b-a0dc-48d1-bd83-d729860fea63

Note: none of the above have `schedVer` (sched V2) set

**Comment by tingsu** [comment]
Thanks for your quick reply and confirmation, @david-allison-1 !  

> Do you have logs from before this happens (I'm specifically looking for how the action for "Undo" action occurred, and where the notes were).

Yes, I have the whole trace. In fact, I tried to maually come up a valid trace. But, up to now, I cannot find a shortest trace. As you said, the main issue is to find the pivot state (e.g., the page with "undo") under which the issue happens. The current trace I provided was cut off by myself from the long trace (the trace is too long...).  One note is that: the event in the trace are all normal UI actions without any hacky events there.  It may be difficult for me to share the log with you, and you will also face the same situation as me (i.e., the prefix trace is too long).

Let me try again to see whether I can come up with a good trace to help fix this issue.

> This happens in the wild as well, so not just an automated testing crash. (Saw a few on 2.8.4 as well)

This crash was found on the recent released version 2.9.6 (let me update this info in the issue to make it more clear).

> Note: none of the above have schedVer (sched V2) set.

I do not quite understand this comment. To be honest, I am not quite familar with Ankidroid's code :)

**Comment by david-allison-1** [comment]
@tingsu Thanks so much for the prompt reply!

> > Note: none of the above have schedVer (sched V2) set.
> 
> I do not quite understand this comment. To be honest, I am not quite familar with Ankidroid's code :)

I was stating to any developers reading that it's unlikely to be an issue with the V2 (experimental) scheduler. Sorry for the confusion.

I'm more than happy to delve through any type of logs if you're comfortable sending them. 

gmail: davidallisongithub

**Comment by tingsu** [comment]
Okay, thanks @david-allison-1 ! Let me prepare the log and send it to you.

**Comment by dae** [comment]
If it's still happening, please let us know.
